### PR TITLE
discourse: update 3.1.0.beta4 -> 3.1.0

### LINKED ISSUE #249901: Build failure: PACKAGENAME
### Steps To Reproduce

1. build discourse
2. `NIXPKGS_ALLOW_INSECURE=1 nix build --impure nixpkgs#discourse`

### Build log

```
error: builder for '/nix/store/wycsqf2smqq19q7gjdsbflk5bgii2rx1-ruby3.2.2-sass-embedded-1.62.0.drv' failed with exit code 1;                                                                     
       last 10 log lines:                                                                                                                                                                        
       > /nix/store/qdpj4g1cjc8chqdwkngckfq6nhwmja4n-ruby3.2.2-sass-embedded-1.62.0/lib/ruby/gems/3.2.0/gems/sass-embedded-1.62.0/ext/sass/Rakefile:21:in `fetch'                                
       > /nix/store/qdpj4g1cjc8chqdwkngckfq6nhwmja4n-ruby3.2.2-sass-embedded-1.62.0/lib/ruby/gems/3.2.0/gems/sass-embedded-1.62.0/ext/sass/Rakefile:21:in `block in <top (required)>'            
       > /nix/store/nr2cs0fw8vlfzchci2ms0hb6srx3hvi9-ruby3.2.2-rake-13.0.6/lib/ruby/gems/3.2.0/gems/rake-13.0.6/exe/rake:27:in `<main>'                                                          
       > Tasks: TOP => default => install => embedded.rb => sass_embedded                                                                                                                        
       > (See full trace by running task with --trace)                                                                                                                                           
       >                                                                                                                                                                                         
       > rake failed, exit code 1                                                                                                                                                                
       >                                                                                                                                                                                         
       > Gem files will remain installed in /nix/store/qdpj4g1cjc8chqdwkngckfq6nhwmja4n-ruby3.2.2-sass-embedded-1.62.0/lib/ruby/gems/3.2.0/gems/sass-embedded-1.62.0 for inspection.             
       > Results logged to /nix/store/qdpj4g1cjc8chqdwkngckfq6nhwmja4n-ruby3.2.2-sass-embedded-1.62.0/lib/ruby/gems/3.2.0/extensions/x86_64-linux/3.2.0/sass-embedded-1.62.0/gem_make.out        
       For full logs, run 'nix log /nix/store/wycsqf2smqq19q7gjdsbflk5bgii2rx1-ruby3.2.2-sass-embedded-1.62.0.drv'.                                                                              
error: 1 dependencies of derivation '/nix/store/7blbl86337qacac9ll75hw7ddg2za2zv-discourse-ruby-env-3.1.0.beta4.drv' failed to build                                                             
error: 1 dependencies of derivation '/nix/store/llsyin7v340acxi6hv2dqr2fgkkf7fms-discourse-3.1.0.beta4.drv' failed to build            
```

### Additional context
gem_make.out
```
current directory: /nix/store/qdpj4g1cjc8chqdwkngckfq6nhwmja4n-ruby3.2.2-sass-embedded-1.62.0/lib/ruby/gems/3.2.0/gems/sass-embedded-1.62.0/ext/sass                                             
/nix/store/cd6pq2f6hb40bv55yd9bzmz0b6gmmc6n-ruby-3.2.2/bin/ruby -rrubygems /nix/store/nr2cs0fw8vlfzchci2ms0hb6srx3hvi9-ruby3.2.2-rake-13.0.6/lib/ruby/gems/3.2.0/gems/rake-13.0.6/exe/rake RUBYAR
CHDIR\=/nix/store/qdpj4g1cjc8chqdwkngckfq6nhwmja4n-ruby3.2.2-sass-embedded-1.62.0/lib/ruby/gems/3.2.0/extensions/x86_64-linux/3.2.0/sass-embedded-1.62.0 RUBYLIBDIR\=/nix/store/qdpj4g1cjc8chqdwk
ngckfq6nhwmja4n-ruby3.2.2-sass-embedded-1.62.0/lib/ruby/gems/3.2.0/extensions/x86_64-linux/3.2.0/sass-embedded-1.62.0                                                                            
rake aborted!                                                                                                                                                                                    
KeyError: key not found: "SASS_EMBEDDED"                                                                                                                                                         
/nix/store/qdpj4g1cjc8chqdwkngckfq6nhwmja4n-ruby3.2.2-sass-embedded-1.62.0/lib/ruby/gems/3.2.0/gems/sass-embedded-1.62.0/ext/sass/Rakefile:21:in `fetch'                                         
/nix/store/qdpj4g1cjc8chqdwkngckfq6nhwmja4n-ruby3.2.2-sass-embedded-1.62.0/lib/ruby/gems/3.2.0/gems/sass-embedded-1.62.0/ext/sass/Rakefile:21:in `block in <top (required)>'                     
/nix/store/nr2cs0fw8vlfzchci2ms0hb6srx3hvi9-ruby3.2.2-rake-13.0.6/lib/ruby/gems/3.2.0/gems/rake-13.0.6/exe/rake:27:in `<main>'                                                                   
Tasks: TOP => default => install => embedded.rb => sass_embedded                                                                                                                                 
(See full trace by running task with --trace)                                                                                                                                                    
                                                                                                                                                                                                 
rake failed, exit code 1                                                                                                                                                                         
```


### Notify maintainers

@talyz 

### Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.45-gnu, NixOS, 23.11 (Tapir), 23.11.20230813.100a155`
 - multi-user?: `yes`

 - sandbox: `yes`
 - version: `nix-env (Nix) 2.15.1`
 - channels(null): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### PR DESCRIPTION
**Pull request by TheNeikos**
Fixes #249901: "discourse Build failure"

I've also built `discourseAllPlugins` without issues

## Description of changes

The `SASS_EMBEDDED` env variable was not visible later in the build. This patch moves it to the preBuild phase, so that it is visible when needed.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Also @Ma27 and @kira-bruneau patched this expression lately... pinging you two here.

**Comment by TheNeikos** [comment]
I have no idea why the variable is not visible later after a `postPatch`. Putting it in the attrset as well as `preBuild` works though...

**Comment by emilylange** [comment]
This is because of #240000 🫠

**Comment by emilylange** [comment]
We should probably bump discourse from `3.1.0.beta4` to `3.1.0`, which has been released in the meantime, first.
It looks to me like the patch (and overwrite) I added for #228716 might no longer be necessary.

**Comment by kira-bruneau** [comment]
> Also @Ma27 and @kira-bruneau patched this expression lately... pinging you two here.

Oh yep, the change I made was to account for building dart-sass-embedded from source.

> This is because of https://github.com/NixOS/nixpkgs/pull/240000 🫠

🫠 oh cool! I like that general approach, but not really sure why it'd break discourse. It also looks like dart-sass-embedded isn't needed anymore either (now that the embedded part was merged into dart-sass)

**Comment by emilylange** [comment]
@TheNeikos would be awesome if you could give updating discourse a try!
See https://github.com/NixOS/nixpkgs/blob/89837aba44c07f4d5c8f9f3601048b199330b44f/pkgs/servers/web-apps/discourse/how_to_update.md

(It's rather difficult, but I am here if you need any help)

Feel free to either re-use this PR or create a new one :)

**Comment by emilylange** [comment]
Are you still interested @TheNeikos?

**Comment by TheNeikos** [comment]
Yes! I did not see that you replied two weeks ago... Lemme get to it

**Comment by TheNeikos** [comment]
PHEW

@emilylange 

Okay, this was a wild ride. The `update.py` script had some assumptions about my system, so I had to fix that first. But today I took the time to do all that.

I followed the `how_to_update` guide, and will test it now on my own discourse instance. And report back

**Comment by TheNeikos** [comment]
image

Looks like it works?

I tried posting, and checked some of the plugins I had. All worked

Critically, I _don't_ have the prometheus plugin, and don't see how test it faithfully.

**Review by emilylange — CHANGES_REQUESTED** [review]
Wow, thank you a lot!

---

Just a few things:

I believe you can drop the whole `dart-sass` section now

```diff
diff --git a/pkgs/servers/web-apps/discourse/default.nix b/pkgs/servers/web-apps/discourse/default.nix
index 23f09f897a6f..b5064cd078ac 100644
--- a/pkgs/servers/web-apps/discourse/default.nix
+++ b/pkgs/servers/web-apps/discourse/default.nix
@@ -38,7 +38,6 @@
 , fixup_yarn_lock
 , nodePackages
 , nodejs_16
-, dart-sass-embedded
 , jq
 , moreutils
 
@@ -189,20 +188,6 @@ let
               cp $(readlink -f ${libpsl}/lib/libpsl.so) vendor/libpsl.x86_64.so
             '';
           };
-          sass-embedded = gems.sass-embedded // {
-            dontBuild = false;
-            # `sass-embedded` depends on `dart-sass-embedded` and tries to
-            # fetch that as `.tar.gz` from GitHub releases. That `.tar.gz`
-            # can also be specified via `SASS_EMBEDDED`. But instead of
-            # compressing our `dart-sass-embedded` just to decompress it
-            # again, we simply patch the Rakefile to symlink that path.
-            patches = [
-              ./rubyEnv/sass-embedded-static.patch
-            ];
-            preBuild = ''
-              export DART_SASS=${dart-sass-embedded}/bin
-            '';
-          };
         };
 
     groups = [
diff --git a/pkgs/servers/web-apps/discourse/rubyEnv/sass-embedded-static.patch b/pkgs/servers/web-apps/discourse/rubyEnv/sass-embedded-static.patch
deleted file mode 100644
index 86c343e30fd1..000000000000
--- a/pkgs/servers/web-apps/discourse/rubyEnv/sass-embedded-static.patch
+++ /dev/null
@@ -1,21 +0,0 @@
-diff --git a/ext/sass/Rakefile b/ext/sass/Rakefile
-index 4ca11d4f82ea..c0450ad6f8f3 100644
---- a/ext/sass/Rakefile
-+++ b/ext/sass/Rakefile
-@@ -18,15 +18,7 @@ file 'protoc.exe' do |t|
- end
- 
- file 'dart-sass' do |t|
--  raise if ENV.key?('DART_SASS')
--
--  gem_install 'sass-embedded', SassConfig.gem_version, SassConfig.gem_platform do |dir|
--    cp_r File.absolute_path("ext/sass/#{t.name}", dir), t.name
--  end
--rescue StandardError
--  archive = fetch(ENV.fetch('DART_SASS') { SassConfig.default_dart_sass })
--  unarchive archive
--  rm archive
-+  symlink(ENV.fetch('DART_SASS'), t.name)
- end
- 
- file 'cli.rb' => %w[dart-sass] do |t|
```

and drop your original commit titled `discourse: fix missing SASS_EMBEDDED during build` (fb267fcbe98823470653a3b62e261628cdbbb1c0).

Commit titles (`current` -> `suggestion`):
- `discourse: Run update script and fix patch for it` -> 
  `discourse: 3.1.0.beta4 -> 3.1.0`
- `discourse: Do not use deprecated field in discourse test` ->
  `nixosTests.discourse: do not use deprecated field`
   (or `nixos/tests/discourse: do not use deprecated field`)
- `discourse: Update discourse plugins` ->
  `discourseAllPlugins: update all`
-  `discourse: Update ABI patch for discourse-prometheus` ->
  `discourseAllPlugins.plugins.discourse-prometheus: update ABI patch`
  (or just keep it, I dunno -- no strong feelings. my suggestion feels a bit ridiculous, but this would be the proper attr name but oh well /shrug)

**Comment by TheNeikos** [comment]
@emilylange I don’t mind removing the patch, but given the comment it still feels applicable? Symlinking seems better than copying and unpacking something from the store.

I’ll fix the commit messages, thanks!

**Comment by emilylange** [comment]
> @emilylange I don’t mind removing the patch, but given the comment it still feels applicable? Symlinking seems better than copying and unpacking something from the store.

That patch is redundant since #240000.
And given the one from #240000 is global, I'd argue we should just remove our local patch :)

**Comment by TheNeikos** [comment]
@emilylange Alright, here we go I think that's all your requests

File: pkgs/servers/web-apps/discourse/rubyEnv/sass-embedded-static.patch
Comment (emilylange) [inline pkgs/servers/web-apps/discourse/rubyEnv/sass-embedded-static.patch:1]: This file is no longer referenced.
I think you forgot to remove it :)

**Review by emilylange — CHANGES_REQUESTED** [review]
Thanks!

---

Result of `nixpkgs-review pr 250735` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>discourse</li>
    <li>discourseAllPlugins</li>
  </ul>
</details>

File: pkgs/servers/web-apps/discourse/rubyEnv/sass-embedded-static.patch
Comment (TheNeikos) [reply pkgs/servers/web-apps/discourse/rubyEnv/sass-embedded-static.patch:1]: Woops, you're right! Sorry about that

**Review by talyz — DISMISSED** [review]
Thanks a lot for working on upgrading discourse!

`nodejs-16.20.2` is marked as insecure, so it needs to be updated to `nodejs_18` to work without `permittedInsecurePackages = ...`. The tests don't pass on ofborg because of this.

This works for me:

```patch
diff --git a/pkgs/servers/web-apps/discourse/default.nix b/pkgs/servers/web-apps/discourse/default.nix
index c86725f1e49b..ed67a36d52fa 100644
--- a/pkgs/servers/web-apps/discourse/default.nix
+++ b/pkgs/servers/web-apps/discourse/default.nix
@@ -37,7 +37,7 @@
 , yarn
 , fixup_yarn_lock
 , nodePackages
-, nodejs_16
+, nodejs_18
 , dart-sass-embedded
 , jq
 , moreutils
@@ -162,9 +162,9 @@ let
                 cd ../..
 
                 mkdir -p vendor/v8/${stdenv.hostPlatform.system}/libv8/obj/
-                ln -s "${nodejs_16.libv8}/lib/libv8.a" vendor/v8/${stdenv.hostPlatform.system}/libv8/obj/libv8_monolith.a
+                ln -s "${nodejs_18.libv8}/lib/libv8.a" vendor/v8/${stdenv.hostPlatform.system}/libv8/obj/libv8_monolith.a
 
-                ln -s ${nodejs_16.libv8}/include vendor/v8/include
+                ln -s ${nodejs_18.libv8}/include vendor/v8/include
 
                 mkdir -p ext/libv8-node
                 echo '--- !ruby/object:Libv8::Node::Location::Vendor {}' >ext/libv8-node/.location.yml
@@ -212,7 +212,7 @@ let
       nodePackages.terser
       nodePackages.patch-package
       yarn
-      nodejs_16
+      nodejs_18
       jq
       moreutils
     ];
```

We usually update to the latest beta release, and `3.2.0.beta1` was just released. That's probably more work than just running the updater, though, so let's do that in a separate PR.

**Comment by TheNeikos** [comment]
@talyz Alrighty! Updated

**Comment by drupol** [comment]
I'm looking forward for this update!

**Review by emilylange — APPROVED** [review]

**Comment by TheNeikos** [comment]
Will this be backported into 23.05 or only available for 23.11 in the future? (not sure what the criteria is for updates like this)

**Comment by matthiasbeyer** [comment]
Added the backport label. IMO this does clearly qualify for a backport, as it updates from a beta release to a stable release.

Lets see whether the bot can backport this.

**Pull request merged**